crates/runc-shim/src/cgroup_memory.rs

```diff
@@ -0,0 +1,128 @@
+use containerd_shim::{
+    error::{Error, Result},
+    io_error, other_error,
+};
+use std::path::Path;
+use tokio::{
+    fs::{self, File},
+    io::{AsyncBufReadExt, AsyncReadExt, BufReader},
+    sync::mpsc::{self, Receiver},
+};
+
+use nix::sys::eventfd;
+use std::os::unix::io::{AsRawFd, FromRawFd};
+
+pub async fn get_path_from_cgorup(pid: u32) -> Result<String> {
+    let proc_path = format!("/proc/{}/cgroup", pid);
+    let file = File::open(proc_path.clone())
+        .await
+        .map_err(io_error!(e, "open {}.", proc_path))?;
+    let mut lines = BufReader::new(file).lines();
+    while let Some(line) =
+        lines
+            .next_line()
+            .await
+            .map_err(io_error!(e, "read file {} failed.", proc_path))?
+    {
+        if line.contains("memory") {
+            let parts: Vec<&str> = line.split(':').collect();
+            if let Some(path) = parts.last() {
+                return Ok(path.to_string());
+            }
+            break;
+        }
+    }
+    Err("Memory cgroup path not found").map_err(other_error!(e, "Error get path:"))?
+}
+
+pub async fn get_existing_cgroup_mem_path(pid_path: String) -> Result<(String, String)> {
+    let mut mountinfo_path = get_path_from_mountinfo().await?;
+    if mountinfo_path.0 == "/" {
+        mountinfo_path.0 = String::from("");
+    }
+    let res = pid_path.trim_start_matches(&mountinfo_path.0).to_string();
+    Ok((res, mountinfo_path.1))
+}
+
+async fn get_path_from_mountinfo() -> Result<(String, String)> {
+    let mountinfo_file = File::open("/proc/self/mountinfo".clone())
+        .await
+        .map_err(io_error!(e, "open {}", "/proc/self/mountinfo"))?;
+
+    // for line in BufReader::new(mountinfo_file).lines() {
+    let mut lines = BufReader::new(mountinfo_file).lines();
+    while let Some(line) = lines
+        .next_line()
+        .await
+        .map_err(io_error!(e, "read file /proc/self/mountinfo failed."))?
+    {
+        if line.contains("cgroup") && line.contains("memory") {
+            return parse_memory_mountroot(&line);
+        }
+    }
+
+    Err("No memory found in mountinfo file.")
+        .map_err(other_error!(e, "Error get path from montinfo:"))?
+}
+
+fn parse_memory_mountroot(line: &str) -> Result<(String, String)> {
```

Comment:
I understand you want to get the memory subsystem cgroup, how about using `CgroupBuilder::new().set_specified_controllers()`? It will give you a `MemoryResorecsBuilder` that implements getter and setter.

Reply:
The reason why i don't use MemoryResorecsBuilder is because cgroup-rs lib doesn't save the third colmn of ‘/proc/self/mountinfo‘, and use it to trim paths get form `/proc/.../cgroup`:  [parse_mountinfo_for_line](https://github.com/kata-containers/cgroups-rs/blob/main/src/hierarchies.rs#L48).
This issue affects the correct retrieval of cgroup paths when running a container whithin a container.
Futuermore, impletention in cgroup-rs is not async when reading files  `/proc/.../cgroup` and `/proc/.../mountinfo`

Reply:
OK. Container in container is not a common scenarios.